### ISSUE TITLE
'volume_size_info failed' error on ZFS datastore

### ISSUE THREAD

**Issue by 2b** [bpg/terraform-provider-proxmox#1027]
**Describe the bug**
Getting an error while creating `proxmox_virtual_environment_vm` resource on ZFS storage with the latest `0.46.3` provider. Resource is created despite the error.

**To Reproduce**

1. Create a resource `proxmox_virtual_environment_vm` from CentOS 8 example from [cloud-image guide](https://github.com/bpg/terraform-provider-proxmox/blob/main/docs/guides/cloud-image.md) using ZFS datastore (fix a qcow2 url prior to that);
2. See error  

```
Error: error get file ENC-STORAGE:vm-102-disk-0 from datastore ENC-STORAGE: received an HTTP 500 response - Reason: volume_size_info on 'ENC-STORAGE:vm-102-disk-0' failed
│
│   with proxmox_virtual_environment_vm.centos_vm,
│   on main.tf line 30, in resource "proxmox_virtual_environment_vm" "centos_vm":
│   30: resource "proxmox_virtual_environment_vm" "centos_vm" {
```

Please also provide a minimal Terraform configuration that reproduces the issue.

```hcl

resource "proxmox_virtual_environment_vm" "centos_vm" {
  name      = "test-centos"
  node_name = local.pve

  disk {
    datastore_id = "ENC-STORAGE"
    file_id      = proxmox_virtual_environment_download_file.centos_cloud_image.id
    interface    = "virtio0"
    iothread     = true
    discard      = "on"
    size         = 20
  }
}

resource "proxmox_virtual_environment_download_file" "centos_cloud_image" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = local.pve
  url          = "https://cloud.centos.org/centos/8-stream/x86_64/images/CentOS-Stream-GenericCloud-8-latest.x86_64.qcow2"
  file_name    = "centos8.img"
}

```

**Expected behavior**
No error.

**Additional context**
There is no error using non-ZFS datastore (e.g. datastore_id = "local").
And no errors using the same ZFS datastore on the version 0.43.0 of the provider, but not with 0.43.1.
So I believe this bug was introduced in 0.43.1.

- Single or clustered Proxmox: clustered
- Provider version (ideally it should be the latest version): 0.46.3
- Terraform version: OpenTofu v1.6.1
- OS (where you run Terraform from):
- Debug logs (`TF_LOG=DEBUG terraform apply`):

**Comment by bpg** [comment]
Good investigation, thank you @2b 

The only related change is in 0.43.1 is #862, and that might've introduced a regression.
Will take a look.

**Comment by bpg** [comment]
Okay, I wasn't able to reproduce it on a simple mirror zpool, there could be something specific to your ZFS configuration at play.


The error is returned from the PVE API request to get the VM volume details. For your environment it will be equivalent of this cli command: 

```
pvesh get /nodes/<node name>/storage/ENC-STORAGE/content/vm-102-disk-0
```

The previous version of the provider was using a slightly different call to get a list of all volumes from the storage in one shot: 
```
pvesh get /nodes/<node name>/storage/ENC-STORAGE/content
```

From the name of your pool I can guess that you're probably using ZFS encryption? This may slow down ZFS management operations, like volumes provisioning. So I think this issues is a manifestation of a race condition, when the provider is trying to get information of a ZFS volume after VM creation, but the volume is not fully provisioned yet and does not have a size.


@2b Could you share some details about your ZFS setup, so we can validate these assumptions?
Is the error always reproducible in your environment?
Would it be possible to add another ZFS storage from a new pool (with default attributes) on your server to see if that makes any difference?

**Comment by 2b** [comment]
Thank you for the detailed response!


> From the name of your pool I can guess that you're probably using ZFS encryption?

Correct. ZFS pool and encrypted dataset was created with these commands:

```sh
zpool create -f -o ashift=12 -o autoreplace=on rpool mirror sdc sdd

zfs create -o encryption=aes-256-gcm -o keyformat=passphrase rpool/encrypted
```

> So I think this issues is a manifestation of a race condition, when the provider is trying to get information of a ZFS volume after VM creation, but the volume is not fully provisioned yet and does not have a size.

I don't think so, applying terraform configuration again throws the same error again while the volume is already provisioned.

> Could you share some details about your ZFS setup, so we can validate these assumptions?
Is the error always reproducible in your environment?

Yes, the error is always reproducible with each version of the provider >= 0.43.1.

A command `pvesh get /nodes/pve-srv2/storage/ENC-STORAGE/content/vm-102-disk-0`

gives the same error: `volume_size_info on 'ENC-STORAGE:vm-102-disk-0' failed`

I've tested the same `pvesh get` command on another PVE cluster with a non-encrypted ZFS dataset, and it works as expected. The PVE version on that cluster is different, though (8.0.3 vs. 7.2-5 on the one with encrypted datasets).

Unfortunately, both are production clusters, and I can't create any new pools, but I've traced the `pvesh` perl script and found it's calling a function from the file `/usr/share/perl5/PVE/Storage/ZFSPoolPlugin.pm` just before throwing an error. So I replaced that file with the on from 8.0.3 version cluster, and `pvesh` doesn't give any error anymore.

Finally, I found [a commit](https://github.com/proxmox/pve-storage/commit/d70d814ccf09968a63755e8443acb12d3132aaad) related to this issue, so this has nothing to do with your provider. Sorry for wasting your time and thank you for such a feature-rich terraform provider!
